src/analyzer/analyzersilence.cpp

```diff
@@ -43,34 +41,64 @@ bool AnalyzerSilence::initialize(TrackPointer pTrack,
     }
 
     m_iFramesProcessed = 0;
-    m_bPrevSilence = true;
     m_iSignalStart = -1;
     m_iSignalEnd = -1;
 
     return true;
 }
 
-bool AnalyzerSilence::processSamples(const CSAMPLE* pIn, const int iLen) {
-    for (int i = 0; i < iLen; i += mixxx::kAnalysisChannels) {
-        // Compute max of channels in this sample frame
-        CSAMPLE fMax = CSAMPLE_ZERO;
-        for (SINT ch = 0; ch < mixxx::kAnalysisChannels; ++ch) {
-            CSAMPLE fAbs = fabs(pIn[i + ch]);
-            fMax = math_max(fMax, fAbs);
+// static
+SINT AnalyzerSilence::findFirstSound(const CSAMPLE* pIn, SINT iLen) {
+    for (SINT i = 0; i < iLen; ++i) {
+        if (fabs(pIn[i]) >= kSilenceThreshold) {
+            return i;
         }
+    }
+    return -1;
+}
 
-        bool bSilence = fMax < m_fThreshold;
-
-        if (m_bPrevSilence && !bSilence) {
-            if (m_iSignalStart < 0) {
-                m_iSignalStart = m_iFramesProcessed + i / mixxx::kAnalysisChannels;
-            }
-        } else if (!m_bPrevSilence && bSilence) {
-            m_iSignalEnd = m_iFramesProcessed + i / mixxx::kAnalysisChannels;
+// static
+SINT AnalyzerSilence::findLastSound(const CSAMPLE* pIn, SINT iLen, SINT firstSound) {
+    DEBUG_ASSERT(firstSound >= -1);
+    SINT lastSound = firstSound;
+    for (SINT i = firstSound + 1; i < iLen; ++i) {
```

Comment:
After splitting the functions `findLastSound()` could now start searching from the end of the buffer in reverse direction and probably exit early.

Why is `firstSound` passed as a parameter instead of adjusting `pIn` and `iLen` when calling this function. This function should do the same as `findFirstSound()`, just in the opposite direction. Then both functions do not need to know about each other and the DEBUG_ASSERT becomes obsolete.

Reply:
> After splitting the functions findLastSound() could now start searching from the end of the buffer in reverse direction and probably exit early.

I did actually consider this as well, but it did not work, because the analyzer is called in chunks and we cannot go backward to a previous chunk.  

> Why is firstSound passed as a parameter instead of adjusting pIn and iLen

Without it we need to add it to pIn and substract it from size and and add it later again. It felt more save to not mess around with the pIn pointer.

Reply:
> A "chunk" should always be a pointer and a length. It keeps the arguments of the functions aligned.

sidenote: perfect usecase for `std::/gsl::span`

Reply:
With iterators the functions could even be combined into one. I just didn't want to open Pandora's box in this review ;)